Add support for unions

jonas opened the pull request.

Commit:
Add support for unions

Commit:
Also test nested inline nested fragment

yawaramin commented on sangria-codegen/src/main/scala/com.mediative.sangria.codegen/Importer.scala:
```diff
@@ -46,6 +47,10 @@ case class Importer(schema: Schema[_, _], document: ast.Document) {
       input.fields.foreach(field => touchType(field.fieldType))
+    case union: UnionType[_] =>
+      types += union
+      union.types.flatMap(_.fields.map(_.fieldType)).foreach(touchType)
```
`foreach` usually returns `()`, you don't need a separate one.

jonas replied:
Done

yawaramin commented on sangria-codegen/src/main/scala/com.mediative.sangria.codegen/Importer.scala:
```diff
@@ -55,42 +60,69 @@ case class Importer(schema: Schema[_, _], document: ast.Document) {
+                }
+
+                Tree.Selection(
+                  Vector(generateField(touch = false, field.outputName, tpe, union = types)))
```
When calling methods using parameter names, it would be better to use them uniformly for all the arguments.

jonas replied:
I agree this is inconsistent. The reason was to make it clear that the first parameter refers to whether or not the type should be "marked for generation". Replaced with explicit calls to `touchType` in 0203832.

yawaramin commented on sangria-codegen/src/main/scala/com.mediative.sangria.codegen/Importer.scala:
```diff
@@ -55,42 +60,69 @@ case class Importer(schema: Schema[_, _], document: ast.Document) {
             }
 
-          case _ =>
+          case None =>
```
It would be slightly better to express this as a `require` assertion at the top of processing, i.e.

```scala
val result = node match {
  case field: ast.Field =>
    require(typeInfo.tpe.isDefined, s"Field without type: $field")

    typeInfo.tpe.get.namedType match {
      ...
    }
```

jonas replied:
Yes, good point. I still have hope to use `Either` in this part of the code eventually.

yawaramin commented on sangria-codegen/src/main/scala/com.mediative.sangria.codegen/ScalametaGenerator.scala:
```diff
@@ -55,8 +55,8 @@ case class ScalametaGenerator(
-  def generateTemplate(traits: Seq[String]): Template = {
-    val ctorNames = traits.map(moduleName + "." + _).map(Ctor.Name.apply)
+  def generateTemplate(traits: Seq[String], prefix: String = moduleName.value + "."): Template = {
+    val ctorNames = traits.map(prefix + _).map(Ctor.Name.apply)
```
The lambda expression can be written as `prefix.+`.

jonas replied:
I find that a bit less readable in this context.

Commit:
Remove resolved FIXME

Commit:
Express expectation that the field has a type using `require`

Commit:
Remove unneeded unit value

Commit:
Document why fragment spreads are resolved explicitly

Commit:
Explicitly mark referenced types at each field generation site

Allows to remove generateField (and its inconsistent use of named
parameters) in favour of simply calling the Tree.Field constructor
directly.
